File: src/api/admin_applications.py
"""Admin endpoints for application (match) management."""

from fastapi import APIRouter, Depends, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.infrastructure.database import get_session
from src.core.infrastructure.dependencies import client_ip, get_current_admin
from src.core.infrastructure.error_handling import service_exception_to_http
from src.core.infrastructure.pagination import DEFAULT_LIMIT, MAX_LIMIT, CursorPage
from src.core.infrastructure.transactions import defer_after_commit, transactional
from src.core.tasks import enqueue_email_task
from src.enums import ApplicationStatus
from src.models import User
from src.schemas import (
    ApplicationNotesUpdate,
    ApplicationRead,
    ApplicationStatusUpdate,
    ApplicationWithDetails,
)
from src.services.applications_admin import (
    delete_application,
    get_application,
    list_applications,
    update_application_notes,
    update_application_status,
)
from src.services.exceptions import ApplicationNotFoundError, InvalidCursorError

router = APIRouter(prefix="/api/admin", tags=["admin"])


@router.get("/applications", response_model=CursorPage[ApplicationWithDetails])
async def get_applications(
    status: ApplicationStatus | None = None,
    job_id: int | None = None,
    candidate_id: int | None = None,
    cursor: str | None = None,
    limit: int = Query(default=DEFAULT_LIMIT, ge=1, le=MAX_LIMIT),
    current_admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> CursorPage[ApplicationWithDetails]:
    """List applications with optional filters, newest first, cursor-paginated."""
    try:
        return await list_applications(
            session,
            status=status,
            job_id=job_id,
            candidate_id=candidate_id,
            cursor=cursor,
            limit=limit,
        )
    except InvalidCursorError as exc:
        raise service_exception_to_http(exc) from exc


@router.get("/applications/{application_id}", response_model=ApplicationWithDetails)
async def get_application_detail(
    application_id: int,
    current_admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> ApplicationWithDetails:
    """Get a single application with full details."""
    try:
        return await get_application(application_id, session)
    except ApplicationNotFoundError as e:
        raise service_exception_to_http(e) from e


@router.put(
    "/applications/{application_id}/status",
    response_model=ApplicationRead,
    status_code=status.HTTP_200_OK,
)
async def update_application_status_endpoint(
    application_id: int,
    body: ApplicationStatusUpdate,
    request: Request,
    current_admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> ApplicationRead:
    """Update application status. Emails (if any) enqueued after commit."""
    try:
        async with transactional(session):
            result, email_payloads = await update_application_status(
                application_id,
                body.status,
                session,
                admin_notes=body.admin_notes,
                actor_user_id=current_admin.id,
                ip_address=client_ip(request),
            )
            for payload in email_payloads:
                defer_after_commit(lambda p=payload: enqueue_email_task(**p))
    except ApplicationNotFoundError as e:
        raise service_exception_to_http(e) from e

    return result


@router.put(
    "/applications/{application_id}/notes",
    response_model=ApplicationRead,
    status_code=status.HTTP_200_OK,
)
async def update_application_notes_endpoint(
    application_id: int,
    body: ApplicationNotesUpdate,
    current_admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> ApplicationRead:
    """Update only the admin_notes field. Does not change status or send email."""
    try:
        async with transactional(session):
            return await update_application_notes(
                application_id, body.admin_notes, session
            )
    except ApplicationNotFoundError as e:
        raise service_exception_to_http(e) from e


@router.delete("/applications/{application_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_application_endpoint(
    application_id: int,
    current_admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> None:
    """Delete an application record. The candidate profile is preserved."""
    try:
        await delete_application(application_id, session)
    except ApplicationNotFoundError as e:
        raise service_exception_to_http(e) from e

File: frontend/src/services/adminApplications.ts
import api from "@/services/api";
import type { CursorPage } from "@/hooks/useInfiniteList";
import type {
  ApplicationRead,
  ApplicationStatus,
  ApplicationStatusUpdate,
  ApplicationWithDetails,
} from "@/types/api";

export interface ApplicationListParams {
  status?: ApplicationStatus;
  job_id?: number;
  candidate_id?: number;
  cursor?: string | null;
  limit?: number;
}

export async function getApplications(
  params?: ApplicationListParams,
  signal?: AbortSignal,
): Promise<CursorPage<ApplicationWithDetails>> {
  const query: Record<string, string | number> = {};
  if (params?.status) query.status = params.status;
  if (params?.job_id != null) query.job_id = params.job_id;
  if (params?.candidate_id != null) query.candidate_id = params.candidate_id;
  if (params?.cursor) query.cursor = params.cursor;
  if (params?.limit != null) query.limit = params.limit;
  const res = await api.get<CursorPage<ApplicationWithDetails>>(
    "/api/admin/applications",
    { params: query, signal },
  );
  return res.data;
}

export async function updateApplicationNotes(
  appId: number,
  adminNotes: string | null,
): Promise<ApplicationRead> {
  const res = await api.put<ApplicationRead>(`/api/admin/applications/${appId}/notes`, {
    admin_notes: adminNotes,
  });
  return res.data;
}

export async function updateApplicationStatus(
  appId: number,
  body: ApplicationStatusUpdate,
): Promise<ApplicationRead> {
  const res = await api.put<ApplicationRead>(
    `/api/admin/applications/${appId}/status`,
    body,
  );
  return res.data;
}

export async function deleteApplication(appId: number): Promise<void> {
  await api.delete(`/api/admin/applications/${appId}`);
}
